internal/controllers/selfhealing/slice.go

```diff
@@ -0,0 +1,188 @@
+)
+
+// sliceController check if the resource slice is deleted but it is still present in the composition current synthesis status.
+// If yes, it will update the composition PendingResynthesis status to trigger re-synthesis process.
+type sliceController struct {
+	client                 client.Client
+	noCacheReader          client.Reader
+	selfHealingGracePeriod time.Duration
+}
+
+func NewSliceController(mgr ctrl.Manager, selfHealingGracePeriod time.Duration) error {
+	s := &sliceController{
+		client:                 mgr.GetClient(),
+		noCacheReader:          mgr.GetAPIReader(),
+		selfHealingGracePeriod: selfHealingGracePeriod,
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("selfHealingSliceController").
+		Watches(&apiv1.Composition{}, newCompositionHandler()).
+		Watches(&apiv1.ResourceSlice{}, newSliceHandler()).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "selfHealingSliceController")).
+		Complete(s)
+}
+
+func (s *sliceController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	comp := &apiv1.Composition{}
+	err := s.client.Get(ctx, req.NamespacedName, comp)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting composition: %w", err)
+	}
+
+	syn := &apiv1.Synthesizer{}
+	syn.Name = comp.Spec.Synthesizer.Name
+	err = s.client.Get(ctx, client.ObjectKeyFromObject(syn), syn)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("gettting synthesizer: %w", err))
+	}
+
+	logger = logger.WithValues("compositionGeneration", comp.Generation,
+		"compositionName", comp.Name,
+		"compositionNamespace", comp.Namespace,
+		"synthesisID", comp.Status.GetCurrentSynthesisUUID())
+
+	// Skip if the composition is not eligible for resynthesis, and check the synthesis result later
+	if notEligibleForResynthesis(comp) {
+		logger.V(1).Info("not eligible for resynthesis when checking the missing resource slice")
+		// Use default grace period
+		if comp.Status.CurrentSynthesis == nil ||
+			comp.Status.CurrentSynthesis.Synthesized == nil ||
+			(s.selfHealingGracePeriod-time.Since(comp.Status.CurrentSynthesis.Synthesized.Time)) <= 0 {
+			return ctrl.Result{Requeue: true, RequeueAfter: s.selfHealingGracePeriod}, nil
+		}
+		return ctrl.Result{Requeue: true, RequeueAfter: s.selfHealingGracePeriod - time.Since(comp.Status.CurrentSynthesis.Synthesized.Time)}, nil
+	}
+
+	// Check if any resource slice referenced by the composition is deleted.
+	for _, ref := range comp.Status.CurrentSynthesis.ResourceSlices {
+		slice := &apiv1.ResourceSlice{}
+		slice.Name = ref.Name
+		slice.Namespace = comp.Namespace
+		err := s.client.Get(ctx, client.ObjectKeyFromObject(slice), slice)
+		if errors.IsNotFound(err) {
+			// Ensure the resource slice is missing by checking the resource from api-server
+			isMissing, err := s.isSliceMissing(ctx, slice)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			if isMissing {
```

Comment:
nit: inverting the logic would make this somewhat easier to read

Reply:
done!